app/sample_manifest_excel/sample_manifest_excel/upload/processor/plate.rb

```diff
@@ -60,6 +62,59 @@ def find_plate_id_for_sample_id(sample_id)
           sample_manifest_asset = SampleManifestAsset.find_by(sanger_sample_id: sample_id)
           sample_manifest_asset.labware&.id
         end
+
+        # A subset of the plate manifests (stock plates not library plates) are required to have
+        # a retention instruction column to describe how they should be disposed of. The column should
+        # have the same value for all manifest rows for the same plate.
+        def check_for_retention_instruction_by_plate
+          retention_error_row, err_msg = non_matching_retention_instructions_for_plates
+          return if retention_error_row.nil?
+
+          errors.add(:base, "Retention instruction checks failed at row: #{retention_error_row.number}. #{err_msg}")
+        end
+
+        # rubocop:disable Metrics/PerceivedComplexity, Metrics/CyclomaticComplexity, Metrics/AbcSize, Metrics/MethodLength
+        def non_matching_retention_instructions_for_plates
+          return nil, nil unless upload.respond_to?(:rows)
+
+          upload
+            .rows
+            .each_with_object({}) do |row, plate_retentions|
+              # ignore empty rows and skip if the retention column is not present
+              if row.columns.blank? || row.data.blank? || row.columns.extract(['retention_instruction']).count.zero?
+                next
+              end
```

Comment:
You could avoid having this `if` check with a `next` operation if you add a `.reject` operation before your `.each_with_object` operation.